### ISSUE TITLE
Running into connection issue; Connection prematurely closed BEFORE response

### ISSUE THREAD

**Issue by pacphi** [Playtika/feign-reactive#64]
I have a UI I've implemented w/ Vaadin Flow that employs `reactive-feign` libs to interact with a backend.  I'm using `Spring Boot 2.1.4.RELEASE` and `Spring Cloud Greenwich.SR1`.

Linking to source and physical deployments, here:

* (backend), https://github.com/pacphi/cf-hoover, https://cf-hoover-hilarious-alligator.apps.pcfone.io/snapshot/summary
* (ui), https://github.com/pacphi/cf-hoover-ui, https://cf-hoover-ui-wacky-crocodile.apps.pcfone.io/

When I visit the UI I'd expect to see one metric rendered, the total number of applications as taken from `/snapshot/summary` endpoint.  But instead we get a fallback to a blank screen.

This is what I see in the logs...
 
```
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at reactor.core.publisher.MonoDelayElement$DelayElementSubscriber.lambda$onNext$0(MonoDelayElement.java:123)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at io.pivotal.cfapp.ui.CfHooverDashboard.lambda$init$0(CfHooverDashboard.java:44)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1[30m2019-04-27T01:05:12,017[m [1;31mERROR [m[[1;34mparallel-1[m] [1;33mr.u.Loggers$Slf4JLogger[m: Scheduler worker in group main failed with an uncaught exception
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1java.lang.NullPointerException
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1505)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at reactor.core.publisher.LambdaMonoSubscriber.onNext(LambdaMonoSubscriber.java:137)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at java.base/java.lang.Thread.run(Unknown Source)
Fri Apr 26 2019
18:05:12.018
[APP/0]
1	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at io.pivotal.cfapp.ui.CfHooverDashboard.lambda$init$0(CfHooverDashboard.java:44)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1[30m2019-04-27T01:05:12,015[m [1;31mERROR [m[[1;34mparallel-1[m] [1;33mr.u.Loggers$Slf4JLogger[m: Operator called default onErrorDropped
Fri Apr 26 2019
18:05:12.016
[APP/0]
1java.lang.NullPointerException
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1505)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at reactor.core.publisher.LambdaMonoSubscriber.onNext(LambdaMonoSubscriber.java:137)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at java.base/java.lang.Thread.run(Unknown Source)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at reactor.core.publisher.MonoDelayElement$DelayElementSubscriber.lambda$onNext$0(MonoDelayElement.java:123)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
Fri Apr 26 2019
18:05:12.016
[APP/0]
1	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27)
Fri Apr 26 2019
18:05:11.927
[APP/0]
1[30m2019-04-27T01:05:11,927[m [32mINFO [m[[1;34mPollingServerListUpdater-0[m] [1;33mc.n.c.ChainedDynamicProperty$ChainLink[m: Flipping property: cf-hoover.ribbon.ActiveConnectionsLimit to use NEXT property: niws.loadbalancer.availabilityFilteringRule.activeConnectionsLimit = 2147483647
Fri Apr 26 2019
18:05:11.498
[APP/0]
1[30m2019-04-27T01:05:11,497[m [33mWARN [m[[1;34mreactor-http-epoll-4[m] [1;33mi.p.c.c.HooverClientFallbackFactory$1[m: Could not obtain snapshot summary. reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
Fri Apr 26 2019
18:05:11.486
[APP/0]
1reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
Fri Apr 26 2019
18:05:11.486
[APP/0]
1[30m2019-04-27T01:05:11,485[m [33mWARN [m[[1;34mreactor-http-epoll-4[m] [1;33mr.u.Loggers$Slf4JLogger[m: [id: 0x0bcd082a, L:/10.255.38.181:50486 ! R:cf-hoover-hilarious-alligator.apps.pcfone.io/10.193.36.220:443] The connection observed an error
Fri Apr 26 2019
18:05:11.018
[APP/0]
1[30m2019-04-27T01:05:11,018[m [36mDEBUG [m[[1;34mhttp-nio-8080-exec-3[m] [1;33mr.c.l.DefaultReactiveLogger[m: [HooverClient#getSummary]--->GET http://cf-hoover-hilarious-alligator.apps.pcfone.io:443/snapshot/summary HTTP/1.1
Fri Apr 26 2019
18:05:10.956
[APP/0]
1[30m2019-04-27T01:05:10,956[m [32mINFO [m[[1;34mhttp-nio-8080-exec-3[m] [1;33mc.n.l.DynamicServerListLoadBalancer[m: DynamicServerListLoadBalancer for client cf-hoover initialized: DynamicServerListLoadBalancer:{NFLoadBalancer:name=cf-hoover,current list of Servers=[cf-hoover-hilarious-alligator.apps.pcfone.io:793032db-2a8b-4e65-5f8c-4f7d],Load balancer stats=Zone stats: {apps.pcfone.io=[Zone:apps.pcfone.io;	Instance count:1;	Active connections count: 0;	Circuit breaker tripped count: 0;	Active connections per server: 0.0;]
Fri Apr 26 2019
18:05:10.956
[APP/0]
1},Server stats: [[Server:cf-hoover-hilarious-alligator.apps.pcfone.io:793032db-2a8b-4e65-5f8c-4f7d; Zone:apps.pcfone.io;	Total Requests:0;	Successive connection failure:0;	Total blackout seconds:0;	Last connection made:Thu Jan 01 00:00:00 UTC 1970;	First connection made: Thu Jan 01 00:00:00 UTC 1970;	Active Connections:0;	total failure count in last (1000) msecs:0;	average resp time:0.0;	90 percentile resp time:0.0;	95 percentile resp time:0.0; min resp time:0.0;	max resp time:0.0;	stddev resp time:0.0]
Fri Apr 26 2019
18:05:10.956
[APP/0]
1]}ServerList:org.springframework.cloud.netflix.ribbon.eureka.DomainExtractingServerList@6304e536
```

This behavior is surprisingly similar to 

* https://github.com/spring-projects/spring-framework/issues/22464
* https://github.com/reactor/reactor-netty/issues/559
* https://github.com/reactor/reactor-netty/issues/498

Just wondering if you observe anything wrong with my implementation?  And if I should perhaps cross-post with the appropriate Spring/Reactor team?  Or is this perhaps a defect in the `reactive-feign` impl?

**Comment by pacphi** [comment]
Nevermind, this article sorted me out, https://committed.software/posts/vaadin-and-spring/vaadin-spring/.

**Comment by pacphi** [comment]
Yeah.  It appears to be an issue with the reactive-feign client.  I stripped out Vaadin dependencies and implemented a [smoke-test](https://github.com/pacphi/cf-hoover-ui/blob/no-vaadin/src/main/java/io/pivotal/cfapp/HooverClientSmokeTest.java) that executes immediately after application startup. It simply logs results of calls via the `@ReactiveFeignClient` annotated [interface](https://github.com/pacphi/cf-hoover-ui/blob/no-vaadin/src/main/java/io/pivotal/cfapp/client/HooverClient.java) and a straight-up `WebClient#get`.  The latter succeeds while the former fails. Not sure what I'm doing incorrect, here.

**Comment by jicui** [comment]
we have the same issue "Connection prematurely closed BEFORE response" ,refer to previous I suspect it is a firewall related or connection pool related ,but actually after I try to enable pool or disable pool ,mode ,neither helps .FireWall return status=0 to client ,which indicate that client (reactor-netty) close the connect in some circumstances 

A "0" (single zero) is logged when the client aborts the connection before the HTTP response was sent. Since no valid HTTP response is sent, Akamai records it with a "0" status code.
Connection prematurely closed BEFORE response

here is the log in client 

```
2019-04-30 02:16:17,319 priority=INFO  app_name=sync thread=reactor-http-epoll-3 location=SaleRepository line=132 Start to call XXXX V1 API URI=XXXXXX```
2019-04-30 02:16:17,321 priority=DEBUG app_name=sync thread=reactor-http-epoll-1 location=PooledConnectionProvider line=254 [id: 0x4ba44386, L:/10.138.194.161:33136 - R:XXXXX/XXXX:443] onStateChange(GET{uri=XXXXX, connection=PooledConnection{channel=[id: 0x4ba44386, L:/10.138.194.161:33136 - R:XXXX/XXXXX:443]}}, [request_sent])```
request_sent
2019-04-30 02:16:17,343 priority=DEBUG app_name=sync thread=reactor-http-epoll-1 location=PooledConnectionProvider line=254 [id: 0x4ba44386, L:/10.138.194.161:33136 ! R:XXXX/XXXX:443] onStateChange(GET{uri=XXXXXXX, connection=PooledConnection{channel=[id: 0x4ba44386, L:/10.138.194.161:33136 ! R:XXXXX/XXXX:443]}}, [response_incomplete])
```
check the timestamp above ,the connection close itself right after he sent request to remote after 20-30ms seems

**Comment by kptfh** [comment]
Hi @pacphi. 
1) Here is the [fix](https://github.com/reactor/reactor-netty/issues/632)  in reactor-netty that should be released soon
2) Can you help me to create simple reproducible test that I can run on my environment?
You can fork [this project](https://github.com/kptfh/feign-reactive-sample) to write the test. It has eureka server, simple web-flux server and feign client already configured.
3) BTW: here is some java.lang.NullPointerException in your log

**Comment by haycco** [comment]
@kptfh anybody solved it? it also happend my services.......

**Comment by kptfh** [comment]
@pacphi @haycco can you check the last released version 1.0.37
It's based on
```
<spring-cloud-bom.version>Greenwich.SR3</spring-cloud-bom.version>
<spring-boot-dependencies-bom.version>2.2.1.RELEASE</spring-boot-dependencies-bom.version>
```

**Comment by lArtiquel** [comment]
Was the issue solved for you guys? @kptfh @pacphi @haycco 

Still occuring for me.

I'm on:
- Reactive Feign Starter version:
`implementation "com.playtika.reactivefeign:feign-reactor-spring-cloud-starter:3.0.3"`
- Spring Cloud 2020.0.3 aka Illford

More on my setup on [Stackoverflow](https://stackoverflow.com/questions/74940081/playtikas-oss-feign-client-org-springframework-web-reactive-function-client-we)

As I understand, the issue should be gone already with the dependency versions I use.
